patches/api/0416-Brigadier-based-command-API.patch

```diff
@@ -0,0 +1,747 @@
++++ b/build.gradle.kts
+@@ -46,7 +46,7 @@ dependencies {
+     implementation("org.ow2.asm:asm:9.4")
+     implementation("org.ow2.asm:asm-commons:9.4")
+     // Paper end
+-
++    api("com.mojang:brigadier:1.0.18") // Paper, expose!
+     api("org.apache.maven:maven-resolver-provider:3.8.5") // Paper, expose
+     compileOnly("org.apache.maven.resolver:maven-resolver-connector-basic:1.7.3")
+     compileOnly("org.apache.maven.resolver:maven-resolver-transport-http:1.7.3")
+diff --git a/src/main/java/io/papermc/paper/command/brigadier/CommandBuilder.java b/src/main/java/io/papermc/paper/command/brigadier/CommandBuilder.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..57a89ca96629a0d80d9d37e45f62ba14d1b2921d
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/command/brigadier/CommandBuilder.java
+@@ -0,0 +1,83 @@
++package io.papermc.paper.command.brigadier;
++
++import com.mojang.brigadier.CommandDispatcher;
++import com.mojang.brigadier.builder.ArgumentBuilder;
++import com.mojang.brigadier.tree.CommandNode;
++import net.kyori.adventure.util.Services;
++import org.bukkit.Bukkit;
++import org.bukkit.plugin.Plugin;
++import org.jetbrains.annotations.ApiStatus;
++import org.jetbrains.annotations.Contract;
++import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
++
++@ApiStatus.NonExtendable
++public abstract class CommandBuilder extends ArgumentBuilder<CommandSourceStack, CommandBuilder> {
++
++    private static final CommandBuilderProvider PROVIDER = Services.service(CommandBuilderProvider.class)
++        .orElseThrow();
++
++    /**
++     * Creates a new command builder.
++     *
++     * @param plugin plugin
++     * @param name   command name
++     * @return builder
++     */
++    @NotNull
++    public static CommandBuilder newCommandBuilder(@NotNull Plugin plugin, @NotNull String name) {
++        return PROVIDER.create(plugin, name);
++    }
++
++    /**
++     * Sets the aliases for this command.
++     *
++     * @param aliases alises
++     * @return self
++     */
++    @NotNull
++    public abstract CommandBuilder aliases(@NotNull String... aliases);
++
++    /**
++     * Gets the description for this command.
++     *
++     * @return description
++     */
++    @Nullable
++    public abstract String getDescription();
++
++    /**
++     * Sets the description for this command.
++     *
++     * @param description description
++     * @return self
++     */
++    @NotNull
++    public abstract CommandBuilder description(@Nullable String description);
++
++    /**
++     * Gets the plugin set for this command builder.
++     *
++     * @return self
++     */
++    @NotNull
++    public abstract Plugin getPlugin();
++
++    /**
++     * Builds the underlying command argument node directly.
++     * @return command argument node
++     * @deprecated It is discouraged from directly using this command node. See {@link CommandBuilder#register()}
++     */
```

Comment:
I think the javadoc shouldn't include anything other than the deprecation notice and a message saying it will throw.